### ISSUE TITLE
Probabilistic Yin and CREPE

### ISSUE THREAD

**Issue by ronggong** [MTG/essentia#788]
As the monophonic pitch extraction algorithms in Essentia are out-of-date, it is appealing to implement two state of the art pitch extraction algorithms which lead to better pitch extraction accuracy:

- [x] Pyin: https://code.soundsoftware.ac.uk/projects/pyin
- [ ] CREPE: https://github.com/marl/crepe

**Comment by dbogdanov** [comment]
For now, we should prioritize PYin over CREPE. Btw, can you provide some links with evaluation of PYin vs Yin?

**Comment by ronggong** [comment]
For the comparison between Yin and Pyin, please check the figure. 3 in the paper: PYIN: A FUNDAMENTAL FREQUENCY ESTIMATOR USING PROBABILISTIC THRESHOLD DISTRIBUTIONS http://matthiasmauch.de/_pdf/mauch_pyin_2014.pdf

**Comment by sankalpg** [comment]
Hello, is there any rough ETA for this feature request ? 
Thanks a lot!

**Comment by dbogdanov** [comment]
@ronggong was working on that, perhaps he can provide some details.

**Comment by ronggong** [comment]
@sankalpg Hi Sankalp, it's not done yet. I will work on it next weekend and try to finish it.

**Comment by dbogdanov** [comment]
Updates? ;)

**Comment by ronggong** [comment]
Pull-request of PYIN sent https://github.com/MTG/essentia/pull/809#issue-240674967

**Comment by ronggong** [comment]
@pabloEntropia what is the status of the tensorflow integration? We need to evaluate if it is still worth to put CREPE into essentia. I guess the criteria should be 
(1) The processing speech of CREPE
(2) Do you think people will be interested in using Essentia version of CREPE than using its original implementation? PYIN has the problem because it is officially only available for Vamp.

**Comment by dbogdanov** [comment]
Yes, I think we should add Crepe too. There's a pending pull request (#802) that adds support for TensorFlow.

**Comment by pabloEntropia** [comment]
Yes, let's do this!  #802 contains a set of algorithms designed to inference from Tensorflow models inside Essentia. Right now we are waiting to have a bunch of models trained on Essentia features before the release. In order to put DL models inside Essentia the steps would be:
1. Train a Tensorflow model as usual. If you need audio features, use Essentia for the extraction. Using the streaming mode would be better because this way you are making sure that the signal flow would be exactly the same on inference time.
2. Ask me how to put the model into the required format as we don't have official documentation for this yet.

**Comment by ronggong** [comment]
@pabloEntropia The issue I want to ask is that is it worth to re-implement CREPE in essentia? My guess is people will go to directly to the original implementation https://github.com/marl/crepe. I can't think of a reason why people wants to install the large essentia instead of just running the small CREPE?

**Comment by ucasiggcas** [comment]
Dear,
Is there any test on Pyin with language C++,
I just want to test the Pyin function, but it was difficult,
Who can help me, please?
Thanks a lot.

**Comment by ucasiggcas** [comment]
Could I use Pyin with Essentia in C++,
and how?
Thanks

**Comment by justinsalamon** [comment]
Hey folks, if you finally do decide to incorporate CREPE into Essentia feel free to hit us up! :)

**Comment by dbogdanov** [comment]
Thanks, @justinsalamon! This is not our priority at the moment, therefore closing this issue for now.

**Comment by dbogdanov** [comment]
Reopening this issue. @palonso, let's review what is missing to add CREPE.

**Comment by justinsalamon** [comment]
In case it's helpful, here are the evaluation results for CREPE vs pYIN:

<img width="469" alt="Screen Shot 2021-03-24 at 1 39 15 PM" src="https://user-images.githubusercontent.com/3009670/112380224-63196300-8ca6-11eb-9f31-49ebaf77904b.png">

<img width="933" alt="Screen Shot 2021-03-24 at 1 39 23 PM" src="https://user-images.githubusercontent.com/3009670/112380248-67458080-8ca6-11eb-8a4c-d28f4c623b40.png">

**Comment by justinsalamon** [comment]
Another important thing to note is that we changed how we estimate the final pitch compared to the paper in two important ways:


## 1. Argmax-local Weighted Averaging

This release of CREPE uses the following weighted averaging formula, which is slightly different from the paper. This only focuses on the neighborhood around the maximum activation, which is shown to further improve the pitch accuracy:

<p align="center"><img src="https://user-images.githubusercontent.com/266841/38990411-68408544-4397-11e8-9e87-ca5a86c5508b.png" width="400"></p>

## 2. Temporal smoothing
By default CREPE does not apply temporal smoothing to the pitch curve, but 
Viterbi smoothing is supported via the optional `--viterbi` command line argument.

**Comment by justinsalamon** [comment]
Finally note that we provide several CREPE models of different sizes so you can choose a desired speed/accuracy tradeoff:

## Model Capacity

CREPE uses the model size that was reported in the paper by default, but can optionally
use a smaller model for computation speed, at the cost of slightly lower accuracy.
You can specify `--model-capacity {tiny|small|medium|large|full}` as the command
line option to select a model with desired capacity.

Here's a benchmark we ran, with the number of parameters in parentheses:

<img width="678" alt="crepe_performance_table" src="https://user-images.githubusercontent.com/3009670/112380859-200bbf80-8ca7-11eb-996a-73c8cf3352c1.png">

Hope this helps!

**Comment by palonso** [comment]
@justinsalamon thanks a lot for the explanation!

Now that TensorFlow inference is fully supported it shouldn't take much effort to integrate CREPE. @dbogdanov I can start working on it.
